include/sys/dbuf.h

```diff
@@ -321,13 +321,18 @@ typedef struct dmu_buf_impl {
 	uint8_t db_dirtycnt;
 } dmu_buf_impl_t;
 
-/* Note: the dbuf hash table is exposed only for the mdb module */
-#define	DBUF_MUTEXES 2048
-#define	DBUF_HASH_MUTEX(h, idx) (&(h)->hash_mutexes[(idx) & (DBUF_MUTEXES-1)])
+typedef struct dbuf_hash_mutex {
+	kmutex_t dhm_mutex ____cacheline_aligned;
+} dbuf_hash_mutex_t;
+
+#define	DBUF_HASH_MUTEX(h, idx) \
+	(&((h)->hash_mutexes[(idx) & ((h)->hash_mutex_mask)]).dhm_mutex)
+
 typedef struct dbuf_hash_table {
 	uint64_t hash_table_mask;
+	uint64_t hash_mutex_mask;
 	dmu_buf_impl_t **hash_table;
-	kmutex_t hash_mutexes[DBUF_MUTEXES] ____cacheline_aligned;
+	dbuf_hash_mutex_t *hash_mutexes ____cacheline_aligned;
```

Comment:
____cacheline_aligned is not useful here anymore now that the array is dynamically allocated.

Keeping this alignment would require overallocating by the cache line size. Then the pointer here could point to an aligned address in the allocation while we keep another pointer with the original address to use when we need to free this at module unload.

Reply:
Also, typical cache line sizes on Intel are 64bytes, but that is not a guarantee. Other architectures especially might have different cache line sizes.

Reply:
Good catch, yes I forgot to drop this.  Fixed.  The comments in `dbuf_init()` also do mention that example calculation is only for x86_64.  You're mileage will vary by architecture.